File: hooks/check_sycophancy.py
#!/usr/bin/env python3
"""
Stop hook: blocks responses that begin with sycophantic agreement phrases.

Sycophancy is a canary for shallow thinking. When Claude reflexively agrees
("you're right", "good point", "great idea"), it usually means it hasn't
evaluated the user's input deeply enough. This hook forces a full rethink,
not just a surface-level rephrasing that removes the trigger phrase.
"""

import json
import re
import sys

SYCOPHANCY_PATTERNS = [
    re.compile(r"you.re right", re.IGNORECASE),
    re.compile(r"you are right", re.IGNORECASE),
    re.compile(r"good point", re.IGNORECASE),
    re.compile(r"great idea", re.IGNORECASE),
    re.compile(r"that makes sense", re.IGNORECASE),
    re.compile(r"good catch", re.IGNORECASE),
    re.compile(r"excellent suggestion", re.IGNORECASE),
    re.compile(r"absolutely right", re.IGNORECASE),
    re.compile(r"that.s a great", re.IGNORECASE),
    re.compile(r"great question", re.IGNORECASE),
    re.compile(r"good question", re.IGNORECASE),
]


def check_sycophancy(message: str) -> str | None:
    snippet = message[:300]
    for pattern in SYCOPHANCY_PATTERNS:
        if pattern.search(snippet):
            return (
                f'Sycophancy detected: "{snippet[:80]}..." — reflexive agreement '
                "is a sign of shallow thinking. Stop, re-examine the user's input "
                "critically, and respond with your honest, independent assessment. "
                "Do not simply remove the agreeable phrase — actually rethink."
            )
    return None


if __name__ == "__main__":
    try:
        parsed = json.loads(sys.stdin.read())
    except (json.JSONDecodeError, ValueError):
        sys.exit(0)

    if parsed.get("stop_hook_active"):
        sys.exit(0)

    message = parsed.get("last_assistant_message", "")
    result = check_sycophancy(message)
    if result:
        print(result, file=sys.stderr)
        sys.exit(2)
    sys.exit(0)

File: hooks/check-sycophancy.ts
#!/usr/bin/env bun

/**
 * Stop hook: blocks responses that begin with sycophantic agreement phrases.
 *
 * Sycophancy is a canary for shallow thinking. When Claude reflexively agrees
 * ("you're right", "good point", "great idea"), it usually means it hasn't
 * evaluated the user's input deeply enough. This hook forces a re-examination.
 */

const SYCOPHANCY_PATTERNS = [
  /you.re right/i,
  /you are right/i,
  /good point/i,
  /great idea/i,
  /that makes sense/i,
  /good catch/i,
  /excellent suggestion/i,
  /absolutely right/i,
  /that.s a great/i,
  /great question/i,
  /good question/i,
] as const

export function checkSycophancy(message: string): string | null {
  const snippet = message.slice(0, 300)
  for (const pattern of SYCOPHANCY_PATTERNS) {
    if (pattern.test(snippet)) {
      return `Sycophancy detected: "${snippet.slice(0, 80)}..." — rephrase without agreement phrases and re-examine whether your reasoning holds up.`
    }
  }
  return null
}

interface HookInput {
  stop_hook_active?: boolean
  last_assistant_message?: string
}

if (import.meta.main) {
  const raw = await Bun.stdin.text()
  let parsed: HookInput
  try {
    parsed = JSON.parse(raw) as HookInput
  } catch {
    process.exit(0)
  }
  if (parsed.stop_hook_active) process.exit(0)

  const message = parsed.last_assistant_message ?? ''
  const result = checkSycophancy(message)
  if (result) {
    console.error(result)
    process.exit(2)
  }
  process.exit(0)
}
